enumerables.rb

```diff
@@ -0,0 +1,137 @@
+      elsif is_a? Range
+        yield to_a[i]
+      end
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    i = 0
+    while i < size
+      yield to_a[i], i
+      i += 1
+    end
+    self
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    result = []
+    to_a.my_each { |e| result << e if yield e }
+    result
+  end
+
+  def my_all?(arg = nil)
+    result = true
+    if block_given?
+      my_each { |e| result = false unless yield e }
+    elsif arg.is_a?(Class) && arg.class != Regexp
+      my_each { |e| result = false unless e.is_a?(arg) }
+    elsif arg.is_a?(Regexp)
+      my_each { |e| result = false unless e.match(arg) }
+    elsif arg.nil?
+      block = proc { |e| e.nil? || e == false }
+      my_each { |_e| result = false if block.call }
+    else
+      block = proc { |e| e == arg }
+      my_each { |_e| result = false unless block.call }
+    end
+    result
+  end
+
+  def my_any?(arg = nil)
+    result = false
+    if block_given?
+      my_each { |e| result = true if yield e }
+    elsif arg.is_a?(Class) && arg.class != Regexp
+      my_each { |e| result = true if e.is_a?(arg) }
+    elsif arg.is_a?(Regexp)
+      my_each { |e| result = true if e.match(arg) }
+    else
+      block = proc { |e| e == arg }
+      my_each { |_e| result = true if block.call }
+    end
+    result
+  end
+
+  def my_none?(arg = nil)
+    result = true
+    if block_given?
+      my_each { |e| result = false if yield e }
+    elsif arg.is_a?(Class) && arg.class != Regexp
+      my_each { |e| result = false if e.is_a?(arg) }
+    elsif arg.is_a?(Regexp)
+      my_each { |e| result = false if e.match(arg) }
+    elsif arg.nil?
+      my_each { |e| result = false if e }
+    else
+      my_each { |e| result = false unless e }
+    end
+    result
+  end
+
+  def my_count(int = nil)
+    count = 0
+    if block_given?
+      my_each { |e| count += 1 if yield e }
+    elsif !int.nil?
+      my_each { |e| count += 1 if e == int }
+    else
+      my_each { |_e| count += 1 }
+    end
+    count
+  end
+
+  def my_map(proc_arg = nil)
+    return to_enum(:my_map) unless block_given? || !proc_arg.nil?
+
+    ary = []
+    if proc_arg.nil?
+      my_each { |e| ary << (yield e) }
+    else
+      my_each { |e| ary << proc_arg.call(e) }
+    end
+    ary
+  end
+
+  def my_inject(arg = nil, sym = nil)
+    result = 1
+    if arg.is_a?(Symbol)
+      result = 1 if arg.to_s.include?('*' || '/')
+      my_each { |e| result = result.public_send arg, e }
+      return result
+    elsif arg.is_a?(Numeric) && sym.is_a?(Symbol)
+      result = arg
+      my_each { |e| result = result.public_send sym, e }
+      return result
+    end
+    result = arg unless arg.nil?
+    if block_given?
+      my_each { |e| result = yield(result, e) }
+    else
+      my_each { |e| result = [result, e] }
+    end
+    result
+  end
+end
```

Comment:
#my_inject raises a "LocalJumpError" when no block or argument is given
#my_inject searches for the longest word in an array of strings
#my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
#my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method